internal/querynodev2/handlers.go

```diff
@@ -249,6 +251,72 @@ func (node *QueryNode) querySegments(ctx context.Context, req *querypb.QueryRequ
 	}, nil
 }
 
+func (node *QueryNode) optimizeSearchParams(ctx context.Context, req *querypb.SearchRequest, channel string) *querypb.SearchRequest {
+	log := log.Ctx(ctx).With(zap.String("channel", channel))
+	// no hook applied, just return
+	if node.queryHook == nil {
+		return req
+	}
+
+	serializedPlan := req.GetReq().GetSerializedExprPlan()
+	// plan not found
+	if serializedPlan == nil {
+		log.Warn("serialized plan not found")
+		return req
+	}
+
+	channelNum := req.GetTotalChannelNum()
+	// not set, change to conservative channel num 1
+	if channelNum <= 0 {
+		channelNum = 1
+	}
+
+	plan := planpb.PlanNode{}
+	err := proto.Unmarshal(serializedPlan, &plan)
+	if err != nil {
+		log.Warn("failed to unmarshal plan", zap.Error(err))
+		return req
+	}
+
+	var segmentNum int
+	delegator, ok := node.delegators.Get(channel)
+	if !ok {
+		log.Warn("delegator not found")
+		return req
+	}
+	sealed, _, version := delegator.GetDistribution().GetCurrent()
+	delegator.GetDistribution().FinishUsage(version)
+	// use shardNum * segments num in shard to estimate total segment number
+	segmentNum = int(len(sealed) * int(channelNum))
```

Comment:
prefer to remove line 281 and modify this to `predictSegmentNum :=`

Reply:
done